crypto/provider_core.c

```diff
@@ -1067,8 +1066,9 @@ static int provider_deactivate(OSSL_PROVIDER *prov, int upcalls,
         return -1;
```

Comment:
It's unclear why we take this lock. Can something else activate it while we deactivate it? And so it's not actually a reference counter, and the reference counter is being abused for something it's not intended?

Note that you're partially reverting #15010.

Reply:
> It's unclear why we take this lock. 

The flag_lock protects access to the flag_activated value which could be changed as a result of this deactivation. I think there is possibly further optimisation that could be done here, because the flag is little more than an indication that the activatecnt is non-zero. We might be able to remove it altogether - but that would require more significant refactoring.

> Can something else activate it while we deactivate it?

Yes

> And so it's not actually a reference counter, and the reference counter is being abused for something it's not intended?

I'm not sure I understand this. This lock is the flag_lock which protects a flag not a reference counter?

> Note that you're partially reverting https://github.com/openssl/openssl/pull/15010.

Yes. This PR is changing the scope of what the flag_lock protects. Previously it protected the activatecnt *and* flag_activated. This PR makes activatecnt an atomic that therefore no longer needs to be protected by the flag_lock.

Reply:
You turned activatecnt into a CRYPTO_REF_COUNT again. If it's a reference count, which I think you're saying it's not, there would be no point in taking a lock, since the only time this functions writes to is it when there is only 1 reference to it, and so nothing else should be able to have a lock on it.

If it's not a reference counter, you should probably use CRYPTO_atomic_add().

Reply:
There are two reference counters, refcnt and activatecnt. One counts normal references and the second counts activation references. I see them both as reference counters. It is possible for activatecnt to drop to zero with refcnt being greater than zero (but not the other way around).

This is similar to the functional and structural references that ENGINEs have.

Reply:
CRYPTO_REF_COUNT is designed for something you free() when the counter reaches 0, and the activation counter is no such thing. I don't know if CRYPTO_REF_COUNT behaves properly for activation counter case on all architectures.

Reply:
Ah...hmmm...it does look like CRYPTO_DOWN_REF has some special handling for the 0 case. I'll look at this again.

Reply:
I changed this to use `CRYPTO_atomic_add` instead.